### PR TITLE
stacks: add deferred to protobuf

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Adds just the protobuf definitions for deferred actions.

Fixes #

## Target Release

1.8.x

### NEW FEATURES

- stacks: add deferred actions to provider protocol

### CODE REVIEW

File: docs/plugin-protocol/tfplugin5.5.proto
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin5.5.proto:1]: You don't need to make any changes here.

File: docs/plugin-protocol/tfplugin6.5.proto
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin6.5.proto:1]: You don't need to make any changes here. What you have in this file should be in tfplugin6.6.proto.

File: docs/plugin-protocol/tfplugin6.6.proto
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin6.6.proto:1]: This is the wrong file here. This shoud be a .proto file and not go.

**Review by liamcervante — COMMENTED** [review]
You should update the symlinked files in https://github.com/hashicorp/terraform/tree/main/internal/tfplugin5 and https://github.com/hashicorp/terraform/tree/main/internal/tfplugin6 so they point to the latest .proto files as well.

**Comment by DanielMSchmidt** [comment]
Oh these files are symlinked, that explains why I edited the wrong files 🤦

File: docs/plugin-protocol/tfplugin6.6.proto [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+        // raw_state is the raw states as stored for the resource.  Core does
+        // not have access to the schema of prior_version, so it's the
+        // provider's responsibility to interpret this value using the
+        // appropriate older schema. The raw_state will be the json encoded
+        // state, or a legacy flat-mapped format.
+        RawState raw_state = 3;
+
+        // deferral_allowed signals that the provider is allowed to defer the
+        // changes. If set the caller needs to handle the deferred response.
+        bool deferral_allowed = 4;
```
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin6.6.proto:368]: Just confirming, we don't need to make this change to UpgradeResourceState

File: docs/plugin-protocol/tfplugin6.6.proto [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+        DynamicValue upgraded_state = 1;
+
+        // diagnostics describes any errors encountered during migration that could not
+        // be safely resolved, and warnings about any possibly-risky assumptions made
+        // in the upgrade process.
+        repeated Diagnostic diagnostics = 2;
+
+        // deferred is set if the provider is deferring the change. If set the caller
+        // needs to handle the deferral.
+        Deferred deferred = 3;
```
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin6.6.proto:383]: Just confirming, we don't need to make this change to UpgradeResourceState

**Review by liamcervante — COMMENTED** [review]
Yeah we don't need to change UpgradeResourceState. I've added comments to the exact places in tfplugin6. You'll need to make equivalent changes to tfplugin5.

File: docs/plugin-protocol/tfplugin6.6.proto [hunk 3]
```diff
@@ -230,6 +230,27 @@ message ServerCapabilities {
...
 
+// Deferred is a message that indicates that change is deferred for a reason.
+message Deferred {
+    // Reason is the reason for deferring the change.
+  enum Reason {
+    // UNKNOWN is the default value, and should not be used.
+    UNKNOWN = 0;
+    // RESOURCE_CONFIG_UNKNOWN is used when parts of the resource configuration
+    // are unknown, e.g. the for_each value is only known after the apply is done.
+    RESOURCE_CONFIG_UNKNOWN = 1;
```
Comment (liamcervante) [inline docs/plugin-protocol/tfplugin6.6.proto:241]: same thing here as well, about the comment.

**Review by liamcervante — APPROVED** [review]
This looks good to me. I've tagged in Brian and Austin in case they want to have a final look / say, but otherwise good to merge. You can probably just merge this when you're back after Easter!

**Comment by austinvalle** [comment]
> This looks good to me. I've tagged in Brian and Austin in case they want to have a final look / say, but otherwise good to merge. You can probably just merge this when you're back after Easter!

@liamcervante & @DanielMSchmidt 

@bflad and I talked with @alexsomesan, @radditude, and @nfagerlund while you both were out. Brian and I want to propose some additional changes to the protocol. (as of now, only additive to what's in this PR, so not sure if that makes it a blocker)

Our main goal is to add some bits to the protocol (specifically `ConfigureProvider` and possibly `GetProviderSchema`) for the `PROVIDER_CONFIG_UNKNOWN` scenario, to protect existing resources that rely on provider configs being fully known to successfully plan.

We should sync up next week when you both get back. (I'm not 100% sure who from Core wants to join so I'll wait to schedule something before playing calendar roulette due to the TZ differences 😆 )

**Comment by DanielMSchmidt** [comment]
@austinvalle Reading through the proposed protocol changes they seem to be only additive to this PR, I'll therefore go ahead and merge this PR and we can do follow up once the discussion settles down on the to be added changes 👍

**Pull request merged**